File: Data Structures & Algorithms/dijkstra/submission-36.cpp
#include <algorithm> 
class Solution {
public:
    unordered_map<int, int> shortestPath(int n, vector<vector<int>>& edges, int src) {
        unordered_map<int, int> distances;
        map<int, vector<std::pair<int, int>>> adj;
        set<int> visiting;
        set<int> visited;
        vector<int> sorted;
            // Min-priority queue
        std::priority_queue<std::pair<int, int>, 
            std::vector<std::pair<int,int>>,
            std::greater<std::pair<int,int>>> min_heap;
        min_heap.push({0, src});
        for (int i =0; i < n; i++){
            adj[i] = std::vector<pair<int, int>>();
        }
        for (vector<int>& edge : edges){
           int s = edge[0];
           int d = edge[1]; 
           int w = edge[2];
           adj[s].push_back({d, w});
        }
        while (!min_heap.empty()){
             std::pair<int, int> cur = min_heap.top();
             min_heap.pop();
             int n1 = cur.second;
             int w1 = cur.first;
             if (distances.find(n1) != distances.end()){
                continue;
             }
             distances[n1] = w1;
             for (std::pair<int,int> e :  adj[n1]){
                int n2 = e.first;
                int w2 = e.second;
                if (distances.find(n2) == distances.end()){
                    min_heap.push({w1 + w2, n2});
                }
             }
        }
        for (int i =0; i < n; i++){
            if (distances.find(i) == distances.end()){
                distances[i] = -1;
            }
        } 
        return distances;
    }
};

File: Data Structures & Algorithms/topologicalSort/submission-36.java
class Solution {
    public List<Integer> topologicalSort(int n, int[][] edges) {
        Map<Integer, List<Integer>> adj = new HashMap<>();
        List<Integer> sorted = new ArrayList<>();
        for (int i = 0; i < n; i++){
            adj.put(i, new ArrayList<>());
        }
        for (int[] edge : edges){
           adj.get(edge[0]).add(edge[1]);
        }
        Set<Integer> visiting = new HashSet<>();
        Set<Integer> visited = new HashSet<>();
        for (int i = 0; i < n; i++){
            if (!dfs(adj, visiting, visited, sorted, i)){
                return new ArrayList<>();
            }
        }
        Collections.reverse(sorted);
        return sorted;
    }

    boolean dfs(Map<Integer, List<Integer>> adj, Set<Integer> visiting,
        Set<Integer> visited, List<Integer> sorted, int src){
        if (visiting.contains(src)){
            return false;
        }
        if (visited.contains(src)){
            return true;
        }
        visiting.add(src);
        for (int neighbor : adj.get(src)){
           if (!dfs(adj, visiting, visited, sorted, neighbor)){
            return false; 
           }
        }
        sorted.add(src);
        visited.add(src);
        visiting.remove(src);
        return true;
    }
}
